src/Features/CSharp/Portable/UseLocalFunction/CSharpUseLocalFunctionDiagnosticAnalyzer.cs

```diff
@@ -358,6 +360,39 @@ private void SyntaxNodeAction(SyntaxNodeAnalysisContext syntaxContext, INamedTyp
             return false;
         }
 
+        private static bool CanReplaceDelegateWithLocalFunction(
+            INamedTypeSymbol delegateType, 
+            LocalDeclarationStatementSyntax localDeclaration,
+            SemanticModel semanticModel,
+            CancellationToken cancellationToken)
+        {
+            var delegateContainingType = delegateType.ContainingType;
+            if (delegateContainingType is null || !delegateContainingType.IsGenericType)
+            {
+                return true;
+            }
+
+            var delegateTypeParamNames = delegateType.GetAllTypeParameters().Select(p => p.Name).ToImmutableHashSet();
+            var localEnclosingSymbol = semanticModel.GetEnclosingSymbol(localDeclaration.SpanStart, cancellationToken);
+            while (localEnclosingSymbol != null)
+            {
+                if (localEnclosingSymbol.Equals(delegateContainingType))
+                {
+                    return true;
+                }
+
+                var typeParams = localEnclosingSymbol.GetTypeParameters();
+                if (typeParams.Any())
+                {
+                    return !typeParams.Any(p => delegateTypeParamNames.Contains(p.Name));
+                }
+
+                localEnclosingSymbol = localEnclosingSymbol.ContainingType;
+            }
+
+            return false;
```

Comment:
I think we should only exit when we reached the type that contains the delegate type. Can this statement be reached? If not, let's add an assertion.

Reply:
➡️ Fixed in 44e67c5796d99ff0dccda995063e0f1092b9cd6f with a new test `TestAvailableIfTypeParameterNotChanged2` added. Also revealed #27950 which is left for future work.